these will fix issue (#14860) and (#15136) for UPDATE_EMAIL feature

saman-jafari opened the pull request:
issuedFor added to token to get client id into the token also redirect uri added to token and then passed to info template for "back to application" functionality

Fixes #14860
Fixes #15136

vmuzikar commented:
@Saman-jafari Thanks for the contribution.

I think this possibly belongs to the @keycloak/core team area.

Saman-jafari commented:
@vmuzikar thanks for the trigger, but I rebased the latest changes from main again, please retrigger the checks thanks a lot.

Saman-jafari commented:
ping!!!
can someone please review this ?

Saman-jafari commented:
@sschu thanks for fast feedback, but it does not add or extend, it fixes the features bug that mentioned in issues mentioned in main comment.
Can you please just take look at it if then needed then I will try to build tests for it thanks a lot for your time

douglaspalmer reviewed (changes requested):
I'm not convinced that the introduction of redirect uri is necessary. VerifyEmailActionToken works as expected and it only sets issuedFor. Can you please add a test for the issue and check if the test passes with only the change to set issuedFor?

Saman-jafari commented:
@douglaspalmer thx for the suggestion and review, yeah I know what'd you mean, but I checked that before committing, and just to clarify yes it works for verify email action because the token generated from the first action that user post has all the info like client and redrectUri but in update email it is one step more first it triggers the action and all is fine to update the email, and it sends an email with the token to user after that it triggers the verify action, but this time in that second token there is no client nor redirecturi, form first action (Update Email) to second action (Verify Email) some information went missing and not transferred among actions, and it makes two issues that you see in main topic above.

If I just set issuedFor redirect URI it does not set the redrect uri and most of the time it just uses account client default URI for redrect back and ignore the URL parameter that user/admin sets in Rest API or client_id and..., I already faced it after I fixed the client when I tried to fix the  #14860, and then it evolved into fixing the redirect in this issue #15136 too.

But still, I could be wrong, if you see that what I'm saying is correct then good for me :) if not, and I missed something in there please point me in the right direction, again I really appreciate that you spend time to review the PR 🙇

douglaspalmer commented:
@Saman-jafari OK, so for now can you please add a test for the correct behaviour and then I'll take another look.

saman-jafari committed:
fix: issuedFor added to token to get client id into the token also redirect uri added to token and then passed to info template for "back to application" functionality
test also added to check the availability of issueFor(azp) and redirect uri in Action
Fixes #14860
Fixes #15136

Saman-jafari commented:
@douglaspalmer so, I wrote/extend the test for the verification token, long weekend from job helped, so it took me almost half a day to get where I want it also reading the code and understanding tests in Keycloak found places that I could extend the admin rest too they are missing all the other actions tests there, there is another case that can be added to UserTest in admin section for admin actions for User-IA anyway, I think I will extend that with another issue that I found after this (issue is when you update email the username stays as old email even if you enabled the username same as email I think I saw someone reported that too as an issue) sorry I let's go back to the main topic :) .
Please check the test, btw I went the whole path again because you mentioned above that issuedFor can be enough, so I wanted to be sure issuedFor can be just enough but sadly when you remove passing reduri to the second action (token Update Email to token Email Verify) the reduri that you delivered with initiated action goes missing and test fails and afterward depends on client config there is no redirect uri, or it is the default one from client if issuedFor provided that is now provided with my fix, but again it is not enough because it is not the reduri that you passed to the action in the first place, or it goes to empty totally, so I must pass it through.


Please check again and let me know about it, thanks for your time appreciated 🙇

Saman-jafari commented:
I'm really sorry for pinging again, but @douglaspalmer ping :( , can you please check the PR and also read my comment along the way, I really appreciate it.

Saman-jafari commented:
@douglaspalmer sorry but ping again

douglaspalmer commented:
@Saman-jafari I still think that adding the field for redirect URI is the wrong solution; we don't do that with the action that are working. I'm looking for an alternative solution.

Saman-jafari commented:
@douglaspalmer appricated again for rejection but I really would like to help here so if you just point me to the right direction that I could not found apparently, I'll appreciate it and will fix the issue :)

douglaspalmer commented:
> @douglaspalmer appricated again for rejection but I really would like to help here so if you just point me to the right direction that I could not found apparently, I'll appreciate it and will fix the issue :)

@Saman-jafari I wasn't rejecting it outright. I played around with your PR. Removing the redirect uri field from your PR still fixes both issues and your test passes. The back to application link is correct in both cases. If you remove the changes related to redirect uri then I think your PR is fine.

Saman-jafari commented:
@douglaspalmer i didn't mean to offend sorry, ok I'll check again and update it's wired that it works without it thx again will ping you really soon

douglaspalmer commented:
@Saman-jafari P.s. The fact that redirect URI isn't substituted into "Back to Application" is a different issue to the correct client ID being used in "Back to Application". You could submit a separate issue for that but I'm not even sure that is a valid use case.

Saman-jafari commented:
@douglaspalmer if you pass redirect URI to required action by admin api it don't take it and you don't see the back to application to correct client, it is use case if you use the redirect URI by admin api

Saman-jafari commented:
@douglaspalmer but you are right I'll remove it and add another issue and maybe there we can fix it :)

douglaspalmer commented:
> @douglaspalmer if you pass redirect URI to required action by admin api it don't take it and you don't see the back to application to correct client, it is use case if you use the redirect URI by admin api

@Saman-jafari I would have expected a redirect uri to be followed instead of showing the info page; I would expect it to replace the back to application link. Which is why I'm not sure that redirect uri is appropriate.

Saman-jafari commented:
@douglaspalmer hmm in theory yes you are right but interestingly none of the required action triggered by Admin API with reduri does that, all just replace the link in back to application, in our project we are using Admin API for some actions and by passing the reduri it will pass it to the Back to Application. Then I think in the new issue we need to extend the test of admin API I was checking it, and it does not have so much of testing for this use cases in it `org.keycloak.testsuite.admin.UserTest`

douglaspalmer commented:
> @douglaspalmer hmm in theory yes you are right but interestingly none of the required action triggered by Admin API with reduri does that, all just replace the link in back to application, in our project we are using Admin API for some actions and by passing the reduri it will pass it to the Back to Application. Then I think in the new issue we need to extend the test of admin API I was checking it, and it does not have so much of testing for this use cases in it `org.keycloak.testsuite.admin.UserTest`

Which actions are you using that replace the "Back to Application" uri? I'll dig deeper.

Saman-jafari commented:
I tested it with UPDATE_PASSWORD, UPDATE_PROFILE, VERIFY_PROFILE, and of course UPDATE_EMAIL

Saman-jafari commented:
@douglaspalmer I ran the test, and it fails if I remove the whole `reduri` but you were right I don't know why I added the redirectUri there as field anyway that is removed now check again please, but the others needed, or the test will fails with reduri not set.
When you run without the whole reduri and build after run :

`
mvn -f testsuite/integration-arquillian/pom.xml -Pauth-server-quarkus-embedded clean install -Dtest=AppInitiatedActionUpdateEmailWithVerificationTest#updateEmailWithRedirect
`
you will get :

`
[ERROR] Failures: 
[ERROR]   AppInitiatedActionUpdateEmailWithVerificationTest.updateEmailWithRedirect:168 
Expected: not null
     but: was null
`

so the passed "reduri" from even client is lost there

Saman-jafari commented:
@douglaspalmer OK my last comment is about the Admin API that we need to investigate or create the new issue for it, as you already mentioned, but now I updated the code without reduri field, so it just uses the client_id to get base uri of client and back to application btn works as expected if you don't want to go through Admin API.

I and I'm sure you will investigate more on the Admin API part and will create another issue for it if no one already did 🙇.

But for these two issue now it should be fixed, both of them I updated the test with check Back to Application link it is working fine now.

@douglaspalmer can you please check if OK then approve ? thx a lot, and again thx for reviewing.

douglaspalmer reviewed (approved):
LGTM

Saman-jafari commented:
@douglaspalmer what is the next step here ? it still need review from the others it says :|

Saman-jafari commented:
@douglaspalmer so what should we do here is there a next step ? Should I do something?

douglaspalmer commented:
> @douglaspalmer so what should we do here is there a next step ? Should I do something?

We just need to wait for a maintainer to approve apply the PR. @keycloak/core-maintainers

pedroigor commented on services/src/main/java/org/keycloak/authentication/actiontoken/updateemail/UpdateEmailActionToken.java:
```diff
@@ -29,10 +29,11 @@ public class UpdateEmailActionToken extends DefaultActionToken {
 		super(userId, TOKEN_TYPE, absoluteExpirationInSecs, null);
 		this.oldEmail = oldEmail;
 		this.newEmail = newEmail;
+		this.issuedFor = clientId;
```
Thinking aloud... It is kinda hard to figure out what makes it work because of how the `issuerFor` is set (when serializing/deserializing the token) and how it is used later to establish the client that originated the action.

pedroigor commented:
@Saman-jafari Thanks! I'll try to backport it if we are going to have another patch release.

deresnick commented:
Sorry - I'm new here.  Can you tell me which version of Keycloak (e.g. 22.0.5) has or will have this fix, or tell me how to periodically check whether it's available in a release?

Saman-jafari commented:
It's alright, no need to be sorry. It has been fixed since 22.0.0. It's in Release Notes too.

deresnick commented:
Thank you!

Best,
David

Nov 9, 2023 17:13:51 SamanJafari ***@***.***>:

> 
> It's alright, no need to be sorry. It has been fixed since 22.0.0. It's in Release Notes too.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/keycloak/keycloak/pull/19110#issuecomment-1804765087], or unsubscribe[https://github.com/notifications/unsubscribe-auth/BD3MRQJNJM655WHAFTSD653YDVIR5AVCNFSM6AAAAAAV7OZCLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMBUG43DKMBYG4].
> You are receiving this because you commented.
> [Tracking image][https://github.com/notifications/beacon/BD3MRQPWIL2QVP6WDTJKDNTYDVIR5A5CNFSM6AAAAAAV7OZCLWWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTLSKDZ6.gif]
>

linas-jusys-surfshark commented:
For me this still happens even if I specify Home URL (baseUrl). Back to Application always points to: https://localhost/{prefix}/realms/{realm}/account/, instead of url specified in baseUrl. Using latest keycloak. But instead of UPDATE_EMAIL, I call for UPDATE_PASSWORD executedAction.

juslintek commented:
For me this still happens even if I specify Home URL (baseUrl). Back to Application always points to: https://localhost/{prefix}/realms/{realm}/account/, instead of url specified in baseUrl. Using latest keycloak. But instead of `UPDATE_EMAIL`, I call for `UPDATE_PASSWORD` executeAction.
